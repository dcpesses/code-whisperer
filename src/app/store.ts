import {configureStore, ThunkAction, Action} from '@reduxjs/toolkit';
import counterReducer from '@/features/redux-counter/counterSlice';
import modalCommandListReducer from '@/features/modal-command-list/modalSlice';
// import twitchReducer from '../features/twitch/twitchSlice';
import channelReducer from '@/features/twitch/channel-slice.js';
import userReducer from '@/features/player-queue/user-slice.js';
import queueReducer from '@/features/player-queue/queue-slice';
import todosReducer from '@/features/todos/todosSlice';

const reducer = {
  counter: counterReducer,
  channel: channelReducer,
  modal: modalCommandListReducer,
  queue: queueReducer,
  user: userReducer,
  todos: todosReducer,
};
export const store = configureStore({ reducer });

export function getStoreWithState(preloadedState?: RootState) {
  return configureStore({ reducer, preloadedState });
}

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
