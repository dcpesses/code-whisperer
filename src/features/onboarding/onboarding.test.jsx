/* eslint-env jest */
import { fireEvent, /*prettyDOM,*/ render, screen, waitFor } from '@testing-library/react';

import { getStoreWithState } from '@/app/store';
import { Provider } from 'react-redux';
import OnboardingOverlay from './index';


describe('OnboardingOverlay', () => {
  let body;
  let btnOptions;

  let store;
  beforeEach(() => {
    store = getStoreWithState({
      onboarding: {
        isOnboarding: false,
        activeStep: 0,
        maxSteps: 2,
      }
    });
    body = (<>Popover body text</>);
    btnOptions = {};
  });
  test('Should render without popover', () => {
    const {container} = render(
      <Provider store={store}>
        <OnboardingOverlay content={body} step={1} btnOptions={btnOptions}>
          Content
        </OnboardingOverlay>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  test('Should render with popover', async() => {
    store.dispatch({ type: 'onboarding/showOnboarding' });
    render(
      <Provider store={store}>
        <OnboardingOverlay content={body} step={1} btnOptions={btnOptions}>
          Content
        </OnboardingOverlay>
      </Provider>
    );
    const popoverElement = await screen.findByRole('tooltip');
    expect(popoverElement).toHaveTextContent('Popover body text');
    expect(popoverElement).toMatchSnapshot();
  });

  test('Should render popover without text or icons', async() => {
    store.dispatch({ type: 'onboarding/showOnboarding' });
    let btnOptionAlt = {
      showIcons: false,
      showText: false
    };
    render(
      <Provider store={store}>
        <OnboardingOverlay content={body} step={1} btnOptions={btnOptionAlt}>
          Content
        </OnboardingOverlay>
      </Provider>
    );
    await screen.findByText('❯');
    const popoverElement = await screen.findByRole('tooltip');
    expect(popoverElement).toHaveTextContent('Popover body text');
    expect(popoverElement).not.toHaveTextContent('Next');
    expect(popoverElement).toHaveTextContent('❯');
    expect(popoverElement).toMatchSnapshot();
  });

  test('Should render with popover and display sequentially', async() => {
    store.dispatch({ type: 'onboarding/showOnboarding' });
    render(
      <Provider store={store}>
        <div>
          <OnboardingOverlay content={(<>First Popover body</>)} step={1} btnOptions={btnOptions}>
            Content
          </OnboardingOverlay>
          <OnboardingOverlay content={(<>Second Popover body</>)} step={2} btnOptions={btnOptions}>
            Content
          </OnboardingOverlay>
        </div>
      </Provider>
    );
    await screen.findByText('First Popover body');
    expect(await screen.findByRole('tooltip')).toMatchSnapshot('initial render');

    fireEvent.click(await screen.findByText('Next'));
    await screen.findByText('Second Popover body');
    expect(await screen.findByRole('tooltip')).toMatchSnapshot('Next btn pressed');

    fireEvent.click(await screen.findByText('Done'));
    expect(await screen.findByRole('tooltip')).toMatchSnapshot('Done btn pressed');
    expect(screen.queryByText('Second Popover body')).toBeNull();
  });

  test('Should render with popover and go back to previous popover', async() => {
    store.dispatch({ type: 'onboarding/showOnboarding' });
    store.dispatch({ type: 'onboarding/showNextStep' });
    render(
      <Provider store={store}>
        <div>
          <OnboardingOverlay content={(<>First Popover body</>)} step={1} btnOptions={btnOptions}>
            Content
          </OnboardingOverlay>
          <OnboardingOverlay content={(<>Second Popover body</>)} step={2} btnOptions={btnOptions}>
            Content
          </OnboardingOverlay>
        </div>
      </Provider>
    );
    await screen.findByText('Second Popover body');
    expect(await screen.findByRole('tooltip')).toMatchSnapshot('initial render');

    fireEvent.click(await screen.findByText('Back'));
    await screen.findByText('First Popover body');
    expect(await screen.findByRole('tooltip')).toMatchSnapshot('Back btn pressed');
  });

  test('Should render with popover and skip over additional popovers', async() => {
    store = getStoreWithState({
      onboarding: {
        isOnboarding: false,
        activeStep: 0,
        maxSteps: 2,
      }
    });
    store.dispatch({ type: 'onboarding/showOnboarding' });
    render(
      <Provider store={store}>
        <div>
          <OnboardingOverlay content={(<>First Popover body</>)} step={1} btnOptions={btnOptions}>
            Content
          </OnboardingOverlay>
          <OnboardingOverlay content={(<>Second Popover body</>)} step={2} btnOptions={btnOptions}>
            Content
          </OnboardingOverlay>
        </div>
      </Provider>
    );
    await screen.findByText('First Popover body');
    expect(await screen.findByRole('tooltip')).toMatchSnapshot('initial render');

    fireEvent.click(await screen.findByTitle('Skip and Close'));
    await waitFor(() => {
      expect(screen.queryByText('First Popover body')).toBeNull();
    });
  });

});

