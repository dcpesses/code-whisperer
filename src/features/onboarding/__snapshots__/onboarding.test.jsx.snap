// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OnboardingOverlay > Should render with popover 1`] = `
<div
  class="fade onboarding-popover show popover bs-popover-top"
  data-popper-escaped="true"
  data-popper-placement="top"
  data-popper-reference-hidden="true"
  role="tooltip"
  style="position: absolute; top: auto; right: auto; bottom: 0px; left: 0px; transform: translate(0px, -8px);"
  x-placement="top"
>
  <div
    class="popover-arrow"
    style="position: absolute; left: 0px; transform: translate(0px, 0px);"
  />
  <h3
    class="popover-header"
  >
    <span>
      <i
        class="bi bi-bookmark-star-fill text-body-secondary"
      />
       
      Step
       
      1
       of 
      2
    </span>
    <button
      aria-label="Close"
      class="btn-close ms-auto p-0"
      title="Skip and Close"
      type="button"
    />
  </h3>
  <div
    class="rounded-bottom raleway-font popover-body"
  >
    Popover body text
    <div
      class="d-flex justify-content-between pt-3"
    >
      <button
        aria-label="Back"
        class="btn btn-secondary btn-sm"
        disabled=""
        type="button"
      >
        <i
          class="bi bi-arrow-left-circle-fill"
        />
         Back
      </button>
      <button
        aria-label="Next"
        class="btn btn-primary btn-sm"
        type="button"
      >
        Next 
        <i
          class="bi bi-arrow-right-circle-fill"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`OnboardingOverlay > Should render with popover and display sequentially > Done btn pressed 1`] = `
<div
  class="fade onboarding-popover popover bs-popover-top"
  data-popper-escaped="true"
  data-popper-placement="top"
  data-popper-reference-hidden="true"
  role="tooltip"
  style="position: absolute; top: auto; right: auto; bottom: 0px; left: 0px; transform: translate(0px, -8px);"
  x-placement="top"
>
  <div
    class="popover-arrow"
    style="position: absolute; left: 0px; transform: translate(0px, 0px);"
  />
  <h3
    class="popover-header"
  >
    <span>
      <i
        class="bi bi-bookmark-star-fill text-body-secondary"
      />
       
      Step
       
      2
       of 
      2
    </span>
    <button
      aria-label="Close"
      class="btn-close ms-auto p-0"
      title="Skip and Close"
      type="button"
    />
  </h3>
  <div
    class="rounded-bottom raleway-font popover-body"
  >
    Second Popover body
    <div
      class="d-flex justify-content-between pt-3"
    >
      <button
        aria-label="Back"
        class="btn btn-secondary btn-sm"
        type="button"
      >
        <i
          class="bi bi-arrow-left-circle-fill"
        />
         Back
      </button>
      <button
        aria-label="Next"
        class="btn btn-primary btn-sm"
        type="button"
      >
        Done 
        <i
          class="bi bi-check-circle-fill"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`OnboardingOverlay > Should render with popover and display sequentially > Next btn pressed 1`] = `
<div
  class="fade onboarding-popover show popover bs-popover-top"
  data-popper-escaped="true"
  data-popper-placement="top"
  data-popper-reference-hidden="true"
  role="tooltip"
  style="position: absolute; top: auto; right: auto; bottom: 0px; left: 0px; transform: translate(0px, -8px);"
  x-placement="top"
>
  <div
    class="popover-arrow"
    style="position: absolute; left: 0px; transform: translate(0px, 0px);"
  />
  <h3
    class="popover-header"
  >
    <span>
      <i
        class="bi bi-bookmark-star-fill text-body-secondary"
      />
       
      Step
       
      2
       of 
      2
    </span>
    <button
      aria-label="Close"
      class="btn-close ms-auto p-0"
      title="Skip and Close"
      type="button"
    />
  </h3>
  <div
    class="rounded-bottom raleway-font popover-body"
  >
    Second Popover body
    <div
      class="d-flex justify-content-between pt-3"
    >
      <button
        aria-label="Back"
        class="btn btn-secondary btn-sm"
        type="button"
      >
        <i
          class="bi bi-arrow-left-circle-fill"
        />
         Back
      </button>
      <button
        aria-label="Next"
        class="btn btn-primary btn-sm"
        type="button"
      >
        Done 
        <i
          class="bi bi-check-circle-fill"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`OnboardingOverlay > Should render with popover and display sequentially > initial render 1`] = `
<div
  class="fade onboarding-popover show popover bs-popover-top"
  data-popper-escaped="true"
  data-popper-placement="top"
  data-popper-reference-hidden="true"
  role="tooltip"
  style="position: absolute; top: auto; right: auto; bottom: 0px; left: 0px; transform: translate(0px, -8px);"
  x-placement="top"
>
  <div
    class="popover-arrow"
    style="position: absolute; left: 0px; transform: translate(0px, 0px);"
  />
  <h3
    class="popover-header"
  >
    <span>
      <i
        class="bi bi-bookmark-star-fill text-body-secondary"
      />
       
      Step
       
      1
       of 
      2
    </span>
    <button
      aria-label="Close"
      class="btn-close ms-auto p-0"
      title="Skip and Close"
      type="button"
    />
  </h3>
  <div
    class="rounded-bottom raleway-font popover-body"
  >
    First Popover body
    <div
      class="d-flex justify-content-between pt-3"
    >
      <button
        aria-label="Back"
        class="btn btn-secondary btn-sm"
        disabled=""
        type="button"
      >
        <i
          class="bi bi-arrow-left-circle-fill"
        />
         Back
      </button>
      <button
        aria-label="Next"
        class="btn btn-primary btn-sm"
        type="button"
      >
        Next 
        <i
          class="bi bi-arrow-right-circle-fill"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`OnboardingOverlay > Should render with popover and go back to previous popover > Back btn pressed 1`] = `
<div
  class="fade onboarding-popover show popover bs-popover-top"
  data-popper-escaped="true"
  data-popper-placement="top"
  data-popper-reference-hidden="true"
  role="tooltip"
  style="position: absolute; top: auto; right: auto; bottom: 0px; left: 0px; transform: translate(0px, -8px);"
  x-placement="top"
>
  <div
    class="popover-arrow"
    style="position: absolute; left: 0px; transform: translate(0px, 0px);"
  />
  <h3
    class="popover-header"
  >
    <span>
      <i
        class="bi bi-bookmark-star-fill text-body-secondary"
      />
       
      Step
       
      1
       of 
      2
    </span>
    <button
      aria-label="Close"
      class="btn-close ms-auto p-0"
      title="Skip and Close"
      type="button"
    />
  </h3>
  <div
    class="rounded-bottom raleway-font popover-body"
  >
    First Popover body
    <div
      class="d-flex justify-content-between pt-3"
    >
      <button
        aria-label="Back"
        class="btn btn-secondary btn-sm"
        disabled=""
        type="button"
      >
        <i
          class="bi bi-arrow-left-circle-fill"
        />
         Back
      </button>
      <button
        aria-label="Next"
        class="btn btn-primary btn-sm"
        type="button"
      >
        Next 
        <i
          class="bi bi-arrow-right-circle-fill"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`OnboardingOverlay > Should render with popover and go back to previous popover > initial render 1`] = `
<div
  class="fade onboarding-popover show popover bs-popover-top"
  data-popper-escaped="true"
  data-popper-placement="top"
  data-popper-reference-hidden="true"
  role="tooltip"
  style="position: absolute; top: auto; right: auto; bottom: 0px; left: 0px; transform: translate(0px, -8px);"
  x-placement="top"
>
  <div
    class="popover-arrow"
    style="position: absolute; left: 0px; transform: translate(0px, 0px);"
  />
  <h3
    class="popover-header"
  >
    <span>
      <i
        class="bi bi-bookmark-star-fill text-body-secondary"
      />
       
      Step
       
      2
       of 
      2
    </span>
    <button
      aria-label="Close"
      class="btn-close ms-auto p-0"
      title="Skip and Close"
      type="button"
    />
  </h3>
  <div
    class="rounded-bottom raleway-font popover-body"
  >
    Second Popover body
    <div
      class="d-flex justify-content-between pt-3"
    >
      <button
        aria-label="Back"
        class="btn btn-secondary btn-sm"
        type="button"
      >
        <i
          class="bi bi-arrow-left-circle-fill"
        />
         Back
      </button>
      <button
        aria-label="Next"
        class="btn btn-primary btn-sm"
        type="button"
      >
        Done 
        <i
          class="bi bi-check-circle-fill"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`OnboardingOverlay > Should render with popover and skip over additional popovers > Skip btn pressed 1`] = `
<div
  class="fade onboarding-popover popover bs-popover-top"
  data-popper-escaped="true"
  data-popper-placement="top"
  data-popper-reference-hidden="true"
  role="tooltip"
  style="position: absolute; top: auto; right: auto; bottom: 0px; left: 0px; transform: translate(0px, -8px);"
  x-placement="top"
>
  <div
    class="popover-arrow"
    style="position: absolute; left: 0px; transform: translate(0px, 0px);"
  />
  <h3
    class="popover-header"
  >
    <span>
      <i
        class="bi bi-bookmark-star-fill text-body-secondary"
      />
       
      Step
       
      1
       of 
      2
    </span>
    <button
      aria-label="Close"
      class="btn-close ms-auto p-0"
      title="Skip and Close"
      type="button"
    />
  </h3>
  <div
    class="rounded-bottom raleway-font popover-body"
  >
    First Popover body
    <div
      class="d-flex justify-content-between pt-3"
    >
      <button
        aria-label="Back"
        class="btn btn-secondary btn-sm"
        disabled=""
        type="button"
      >
        <i
          class="bi bi-arrow-left-circle-fill"
        />
         Back
      </button>
      <button
        aria-label="Next"
        class="btn btn-primary btn-sm"
        type="button"
      >
        Next 
        <i
          class="bi bi-arrow-right-circle-fill"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`OnboardingOverlay > Should render with popover and skip over additional popovers > initial render 1`] = `
<div
  class="fade onboarding-popover show popover bs-popover-top"
  data-popper-escaped="true"
  data-popper-placement="top"
  data-popper-reference-hidden="true"
  role="tooltip"
  style="position: absolute; top: auto; right: auto; bottom: 0px; left: 0px; transform: translate(0px, -8px);"
  x-placement="top"
>
  <div
    class="popover-arrow"
    style="position: absolute; left: 0px; transform: translate(0px, 0px);"
  />
  <h3
    class="popover-header"
  >
    <span>
      <i
        class="bi bi-bookmark-star-fill text-body-secondary"
      />
       
      Step
       
      1
       of 
      2
    </span>
    <button
      aria-label="Close"
      class="btn-close ms-auto p-0"
      title="Skip and Close"
      type="button"
    />
  </h3>
  <div
    class="rounded-bottom raleway-font popover-body"
  >
    First Popover body
    <div
      class="d-flex justify-content-between pt-3"
    >
      <button
        aria-label="Back"
        class="btn btn-secondary btn-sm"
        disabled=""
        type="button"
      >
        <i
          class="bi bi-arrow-left-circle-fill"
        />
         Back
      </button>
      <button
        aria-label="Next"
        class="btn btn-primary btn-sm"
        type="button"
      >
        Next 
        <i
          class="bi bi-arrow-right-circle-fill"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`OnboardingOverlay > Should render without popover 1`] = `
<div>
  <div>
    Content
  </div>
</div>
`;
